metals/src/main/scala/scala/meta/internal/metals/FormattingProvider.scala

```diff
@@ -215,6 +223,155 @@ final class FormattingProvider(
+    inspectDialectRewrite(initialConfig) match {
+      case Some(rewrite) => rewrite.rewrite(versionText)
+      case None => versionText
+    }
+  }
+
+  private def inspectDialectRewrite(
+      config: ScalafmtConfig
+  ): Option[DialectRewrite] = {
+
+    import ScalafmtDialect.ord
+
+    def nonExcludedSources(sourceItem: AbsolutePath): List[AbsolutePath] =
+      if (sourceItem.exists)
+        FileIO
+          .listAllFilesRecursively(sourceItem)
+          .filter(f => f.isScalaFilename && !config.isExcluded(f))
+          .toList
+      else
+        List.empty
+
+    def hasCorrectDialectOverride(
+        files: List[AbsolutePath],
+        dialect: ScalafmtDialect
+    ): Boolean = {
+      config.fileOverrides.nonEmpty &&
+      files.forall(p => config.overrideFor(p).exists(d => ord.gteq(dialect, d)))
+    }
+
+    def inferDialectForSourceItem(
+        sourceItem: AbsolutePath,
+        buildTargetIds: List[BuildTargetIdentifier],
+        default: ScalafmtDialect
+    ): Option[ScalafmtDialect] = {
+      if (buildTargetIds.nonEmpty && sourceItem.exists) {
+        val required =
+          buildTargetIds
+            .flatMap(buildTargets.scalaTarget)
+            .map(_.fmtDialect)
+            .sorted
+            .headOption
+            .getOrElse(ScalafmtDialect.Scala213)
+
+        if (ord.gt(required, default)) {
+          nonExcludedSources(sourceItem) match {
+            case Nil => None
+            case items if !hasCorrectDialectOverride(items, required) =>
+              Some(required)
+            case _ => None
+          }
+        } else None
+      } else None
+    }
+
+    val default = config.runnerDialect.getOrElse(ScalafmtDialect.Scala213)
+
+    val itemsRequiresUpgrade =
+      buildTargets.sourceItemsToBuildTargets.toList.flatMap {
+        case (path, ids) =>
+          inferDialectForSourceItem(path, ids.asScala.toList, default).map(d =>
+            (path, d)
+          )
+      }
+
+    if (itemsRequiresUpgrade.nonEmpty) {
+      val (items, dialects) = itemsRequiresUpgrade.unzip
+      val directories = items.map(_.toRelative(workspace))
+
+      val nonSbtTargets = buildTargets.all.toList.filter(!_.isSbt)
+      val needFileOverride =
+        nonSbtTargets.map(_.fmtDialect).distinct.size > 1
+      val upgradeType =
+        if (needFileOverride && config.fileOverrides.nonEmpty)
+          RewriteType.Manual
+        else if (needFileOverride)
+          RewriteType.FileOverrideDialect
+        else
+          RewriteType.GlobalDialect
+
+      val out = DialectRewrite(
+        directories,
+        dialects.max,
+        upgradeType,
+        config
+      )
+      Some(out)
+    } else None
+  }
+
+  private def checkIfDialectUpgradeRequired(
+      config: ScalafmtConfig
+  ): Future[Unit] = {
+
+    if (tables.dismissedNotifications.UpdateScalafmtConf.isDismissed)
+      Future.unit
+    else {
+      inspectDialectRewrite(config) match {
```

Comment:
One last think I would do is maybe put this into a Future, so that we don't block on the check.

Reply:
```suggestion
      Future(inspectDialectRewrite(config)).flatMap{
```